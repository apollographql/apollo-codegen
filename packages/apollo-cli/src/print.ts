import {
  GraphQLSchema,
  TypeMap,
  GraphQLObjectType,
  print,
  astFromValue,
} from "graphql";

enum SEVERITY {
  BREAKING,
  INFO,
}
enum CHANGES {
  TYPE_REMOVED,
  TYPE_ADDED,
}

const levels = {
  [SEVERITY.BREAKING]: [],
  [SEVERITY.INFO]: [],
};

const isInvalid = value => value === undefined || value !== value;

export const findRemovedTypes = (current: TypeMap, next: TypeMap) => {
  const messages = {
    [CHANGES.TYPE_REMOVED]: { ...levels },
    [CHANGES.TYPE_ADDED]: { ...levels },
  };
  // current => next
  Object.keys(current).forEach(typeName => {
    const type = current[typeName];
    // if we have removed this type we can early exit since thats the
    // most critical information
    if (!next[typeName]) {
      messages[CHANGES.TYPE_REMOVED][SEVERITY.BREAKING].push({
        type,
        code: CHANGES.TYPE_REMOVED,
        description: `${typeName} was removed`,
        printed: `type ${typeName}${getObjectNameFromType(type)} {}`,
      });
      return;
    }
  });

  // next => current
  Object.keys(next).forEach(typeName => {
    // if we have a new type we can early exit since thats the
    // most critical information
    const type = next[typeName];
    if (!current[typeName]) {
      messages[CHANGES.TYPE_ADDED][SEVERITY.INFO].push({
        type,
        code: CHANGES.TYPE_ADDED,
        description: `${typeName} was added`,
        printed: `type ${typeName}${getObjectNameFromType(type)} {
${printFields(type)}
}`,
      });
      return;
    }
  });
  return messages;
};

const printInputValue = arg => {
  let argDecl = `${arg.name}: ${String(arg.type)}`;
  if (!isInvalid(arg.defaultValue)) {
    argDecl += ` = ${print(astFromValue(arg.defaultValue, arg.type))}`;
  }
  return argDecl;
};

const printArgs = (args, indentation = "") => {
  if (args.length === 0) return "";
  return `(${args.map(printInputValue).join(", ")})`;
};

const printFields = type =>
  Object.values(type.getFields())
    .map((f, i) => `  ${f.name}${printArgs(f.args, "  ")}: ${String(f.type)}`)
    .join("\n");

const getObjectNameFromType = (type: GraphQLObjectType): string => {
  const interfaces = type.getInterfaces();
  return interfaces.length
    ? " implements " + interfaces.map(i => i.name).join(" & ")
    : "";
};

const headers = {
  breaking: val => (val.length > 0 ? "# Breaking" : ""),
  info: val => (val.length > 0 ? "# Safe" : ""),
};

export const printWithChanges = (current, next) => {
  const currentTypeMap = current.getTypeMap();
  const newTypeMap = next.getTypeMap();

  const details = findRemovedTypes(currentTypeMap, newTypeMap);
  return `
${headers.breaking(details[CHANGES.TYPE_REMOVED][SEVERITY.BREAKING])}
${details[CHANGES.TYPE_REMOVED][SEVERITY.BREAKING]
    .map(({ printed }) => printed)
    .join("\n")}

${headers.info(details[CHANGES.TYPE_REMOVED][SEVERITY.INFO])}
${details[CHANGES.TYPE_REMOVED][SEVERITY.INFO]
    .map(({ printed }) => printed)
    .join("\n")}
  `;
};
