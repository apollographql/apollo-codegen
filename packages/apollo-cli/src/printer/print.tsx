import { uniqBy } from "lodash";
import * as React from "react";
import { renderToStaticMarkup } from "react-dom/server";
import {
  print,
  GraphQLSchema,
  FieldDefinitionNode,
  EnumValueDefinitionNode,
  ObjectTypeDefinitionNode,
  InputObjectTypeDefinitionNode,
  InterfaceTypeDefinitionNode,
  EnumTypeDefinitionNode,
  UnionTypeDefinitionNode
} from "graphql";

import {
  ChangeType,
  Change,
  TypeKind,
  DiffField,
  DiffEnum
} from "./ast";
import * as decode from "decode-html";

import { diffSchemas } from "./diff";

import { f, w, n } from "./emoji";

const Field: React.SFC<{ field: FieldDefinitionNode }> = ({ field }) => (
  <>{print(field)}</>
);
const Value: React.SFC<{ value: EnumValueDefinitionNode }> = ({ value }) => (
  <>{print(value)}</>
);

const nameFromKind = (kind: string) => {
  switch (kind) {
    case TypeKind.ENUM:
      return "enum";
    case TypeKind.OBJECT:
      return "type";
    case TypeKind.INPUT_OBJECT:
      return "input";
    case TypeKind.SCALAR:
      return "scalar";
    case TypeKind.INTERFACE:
      return "interface";
    case TypeKind.UNION:
      return "union";
    default:
      return null;
  }
};
const Fields: React.SFC<{
  name: string;
  fields: FieldDefinitionNode[];
}> = ({ name, fields }) =>
  fields.length > 0 ? (
    <>
      {"\n"}
      {"  "}
      <Header name={name} />
      {fields.map((field, i) => (
        <React.Fragment key={i}>
          {"  "}
          <Field field={field} key={i} />
          {"\n"}
        </React.Fragment>
      ))}
    </>
  ) : null;

const Values: React.SFC<{
  name: string;
  values: EnumValueDefinitionNode[];
}> = ({ name, values }) =>
  values.length > 0 ? (
    <>
      {"\n"}
      {"  "}
      <Header name={name} />
      {values.map((value, i) => (
        <React.Fragment key={i}>
          {"  "}
          <Value value={value} key={i} />
          {"\n"}
        </React.Fragment>
      ))}
    </>
  ) : null;

const removals = ["TYPE_REMOVED"];
const Type: React.SFC<{ change: Change }> = ({ change }) => {
  const type = change.type;
  if (!type) return null;
  const typeName = nameFromKind(type.kind);
  switch (type.kind) {
    case TypeKind.UNION:
      const t = change.type as UnionTypeDefinitionNode;
      if (!t.types) return null;
      const types = t.types.map(({ name }) => name.value);
      return (
        <>
          {typeName} {type.name.value} = {types.join(" | ")}
          {"\n"}
        </>
      );
    case TypeKind.SCALAR:
      return (
        <>
          {typeName} {type.name.value}
          {"\n"}
        </>
      );
    case TypeKind.ENUM: {
      const t = type as EnumTypeDefinitionNode;

      const values = t.values as DiffEnum[];

      const failure = values.filter(
        ({ change }) => change && change.change === ChangeType.FAILURE
      );
      const warning = values.filter(
        ({ change }) => change && change.change === ChangeType.WARNING
      );
      const notice = values.filter(
        ({ change }) => change && change.change === ChangeType.NOTICE
      );

      return (
        <>
          {typeName} {type.name.value}
          {" { "}
          {removals.includes(change.code) ? null : change.code ===
          "TYPE_ADDED" ? (
            <>
              {"\n  "}
              <Values name={`${n} Notice ${n}`} values={values} />
              {"\n"}
            </>
          ) : (
            <>
              {"\n  "}
              <Values name={`${f} Failure ${f}`} values={failure} />
              <Values name={`${w} Warning ${w}`} values={warning} />
              <Values name={`${n} Notice ${n}`} values={notice} />
              {"\n"}
            </>
          )}
          {"}\n"}
        </>
      );
    }
    case TypeKind.INTERFACE:
    case TypeKind.OBJECT:
    case TypeKind.INPUT_OBJECT: {
      const t = type as
        | ObjectTypeDefinitionNode
        | InterfaceTypeDefinitionNode
        | InputObjectTypeDefinitionNode;

      const fields: DiffField[] = t.fields
        ? (t as ObjectTypeDefinitionNode).fields!.filter(
            ({ change }: DiffField) => change
          )
        : [];

      const failure = fields.filter(
        ({ change }) => change!.change === ChangeType.FAILURE
      );
      const warning = fields.filter(
        ({ change }) => change!.change === ChangeType.WARNING
      );
      const notice = fields.filter(
        ({ change }) => change!.change === ChangeType.NOTICE
      );
      const interfaces = (t as ObjectTypeDefinitionNode).interfaces
        ? (t as ObjectTypeDefinitionNode).interfaces!
        : [];
      const implementedInterfaces = interfaces.length
        ? " implements " +
          interfaces.map(({ name }: any) => name.value).join(" & ")
        : "";

      return (
        <>
          {typeName} {t.name.value}
          {implementedInterfaces}
          {" { "}
          {removals.includes(change.code) ||
          (change.code !== "TYPE_ADDED" &&
            fields.length === 0) ? null : change.code === "TYPE_ADDED" ? (
            <>
              {"\n  "}
              <Fields
                name={`${n} Notice ${n}`}
                fields={t.fields as DiffField[]}
              />
              {"\n"}
            </>
          ) : (
            <>
              {"\n  "}
              <Fields name={`${f} Failure ${f}`} fields={failure} />
              <Fields name={`${w} Warning ${w}`} fields={warning} />
              <Fields name={`${n} Notice ${n}`} fields={notice} />
              {"\n"}
            </>
          )}
          {"}\n"}
        </>
      );
    }
    default:
      return null;
  }
};

const Header: React.SFC<{ name: string }> = ({ name }) => (
  <>{"# " + name + "\n"}</>
);

const Schema: React.SFC<{ changes: Change[] }> = ({ changes }) => {
  const filteredChanges = uniqBy(changes.reverse(), "type.name.value");
  return (
    <>
      {filteredChanges.map((change, i) => (
        <React.Fragment key={i}>
          <Type change={change} />
          {"\n"}
        </React.Fragment>
      ))}
    </>
  );
};

export const printChanges = (changes: Change[]) =>
  decode(renderToStaticMarkup(<Schema changes={changes} />));

export const printFromSchemas = (current: GraphQLSchema, next: GraphQLSchema) =>
  printChanges(diffSchemas(current.getTypeMap(), next.getTypeMap()));
