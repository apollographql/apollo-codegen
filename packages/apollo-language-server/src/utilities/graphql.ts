import {
  GraphQLSchema,
  GraphQLCompositeType,
  GraphQLObjectType,
  GraphQLInterfaceType,
  GraphQLUnionType,
  GraphQLField,
  FieldNode,
  SchemaMetaFieldDef,
  TypeMetaFieldDef,
  TypeNameMetaFieldDef,
  ASTNode,
  Kind,
  NameNode,
  TypeSystemDefinitionNode,
  TypeSystemExtensionNode
} from "graphql";

export function isNode(maybeNode: any): maybeNode is ASTNode {
  return maybeNode && typeof maybeNode.kind === "string";
}

export type NamedNode = ASTNode & {
  name: NameNode;
};

export function isNamedNode(node: ASTNode): node is NamedNode {
  return "name" in node;
}

export function isTypeSystemDefinitionNode(
  node: ASTNode
): node is TypeSystemDefinitionNode {
  switch (node.kind) {
    case Kind.SCHEMA_DEFINITION:
    case Kind.SCALAR_TYPE_DEFINITION:
    case Kind.OBJECT_TYPE_DEFINITION:
    case Kind.INTERFACE_TYPE_DEFINITION:
    case Kind.UNION_TYPE_DEFINITION:
    case Kind.ENUM_TYPE_DEFINITION:
    case Kind.INPUT_OBJECT_TYPE_DEFINITION:
    case Kind.DIRECTIVE_DEFINITION:
      return true;
    default:
      return false;
  }
}

export function isTypeSystemExtensionNode(
  node: ASTNode
): node is TypeSystemExtensionNode {
  switch (node.kind) {
    // case Kind.SCHEMA_EXTENSION:
    case Kind.SCALAR_TYPE_EXTENSION:
    case Kind.OBJECT_TYPE_EXTENSION:
    case Kind.INTERFACE_TYPE_EXTENSION:
    case Kind.UNION_TYPE_EXTENSION:
    case Kind.ENUM_TYPE_EXTENSION:
    case Kind.INPUT_OBJECT_TYPE_EXTENSION:
      return true;
    default:
      return false;
  }
}

export function highlightNodeForNode(node: ASTNode): ASTNode {
  switch (node.kind) {
    case Kind.VARIABLE_DEFINITION:
      return node.variable;
    default:
      return isNamedNode(node) ? node.name : node;
  }
}

/**
 * Not exactly the same as the executor's definition of getFieldDef, in this
 * statically evaluated environment we do not always have an Object type,
 * and need to handle Interface and Union types.
 */
export function getFieldDef(
  schema: GraphQLSchema,
  parentType: GraphQLCompositeType,
  fieldAST: FieldNode
): GraphQLField<any, any> | undefined {
  const name = fieldAST.name.value;
  if (
    name === SchemaMetaFieldDef.name &&
    schema.getQueryType() === parentType
  ) {
    return SchemaMetaFieldDef;
  }
  if (name === TypeMetaFieldDef.name && schema.getQueryType() === parentType) {
    return TypeMetaFieldDef;
  }
  if (
    name === TypeNameMetaFieldDef.name &&
    (parentType instanceof GraphQLObjectType ||
      parentType instanceof GraphQLInterfaceType ||
      parentType instanceof GraphQLUnionType)
  ) {
    return TypeNameMetaFieldDef;
  }
  if (
    parentType instanceof GraphQLObjectType ||
    parentType instanceof GraphQLInterfaceType
  ) {
    return parentType.getFields()[name];
  }

  return undefined;
}
