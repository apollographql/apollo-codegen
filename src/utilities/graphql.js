// Utility functions extracted from graphql-js

import {
  SchemaMetaFieldDef,
  TypeMetaFieldDef,
  TypeNameMetaFieldDef,
  GraphQLObjectType,
  GraphQLInterfaceType,
  GraphQLUnionType
} from 'graphql';

/**
 * Extracts the root type of the operation from the schema.
 */
export function getOperationRootType(schema, operation) {
  switch (operation.operation) {
    case 'query':
      return schema.getQueryType();
    case 'mutation':
      const mutationType = schema.getMutationType();
      if (!mutationType) {
        throw new GraphQLError(
          'Schema is not configured for mutations',
          [operation]
        );
      }
      return mutationType;
    case 'subscription':
      const subscriptionType = schema.getSubscriptionType();
      if (!subscriptionType) {
        throw new GraphQLError(
          'Schema is not configured for subscriptions',
          [operation]
        );
      }
      return subscriptionType;
    default:
      throw new GraphQLError(
        'Can only compile queries, mutations and subscriptions',
        [operation]
      );
  }
}

/**
 * Not exactly the same as the executor's definition of getFieldDef, in this
 * statically evaluated environment we do not always have an Object type,
 * and need to handle Interface and Union types.
 */
export function getFieldDef(schema, parentType, fieldAST) {
  const name = fieldAST.name.value;
  if (name === SchemaMetaFieldDef.name &&
      schema.getQueryType() === parentType) {
    return SchemaMetaFieldDef;
  }
  if (name === TypeMetaFieldDef.name &&
      schema.getQueryType() === parentType) {
    return TypeMetaFieldDef;
  }
  if (name === TypeNameMetaFieldDef.name &&
      (parentType instanceof GraphQLObjectType ||
       parentType instanceof GraphQLInterfaceType ||
       parentType instanceof GraphQLUnionType)
  ) {
    return TypeNameMetaFieldDef;
  }
  if (parentType instanceof GraphQLObjectType ||
      parentType instanceof GraphQLInterfaceType) {
    return parentType.getFields()[name];
  }
}
