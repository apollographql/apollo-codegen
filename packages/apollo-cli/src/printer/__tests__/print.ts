import * as fs from "fs";
import * as path from "path";
import { buildSchema, parse } from "graphql";
const gql = String.raw;

import { printFromSchemas, printChanges } from "../print";
import { diffSchemas } from "../diff";

const initial = fs.readFileSync(
  path.join(
    __dirname,
    "../../commands/schema/__tests__/fixtures/schema.graphql"
  ),
  { encoding: "utf8" }
);
const change = fs.readFileSync(
  path.join(__dirname, "../../commands/schema/__tests__/fixtures/next.graphql"),
  { encoding: "utf8" }
);

const schemas = (sd1, sd2) => ({
  current: buildSchema(sd1),
  next: buildSchema(sd2),
});

const compare = (name, sd1, sd2, debug = false) => {
  it(name, () => {
    const { current, next } = schemas(sd1, sd2);
    const sdl = printFromSchemas(current, next);
    if (debug) console.log("DEBUG\n" + sdl);
    if (!debug) expect(sdl).toMatchSnapshot();
  });
};

const fcompare = (name, sd1, sd2, debug = true) => {
  fit(name, () => {
    const { current, next } = schemas(sd1, sd2);
    const sdl = printFromSchemas(current, next);
    if (debug) console.log("DEBUG\n" + sdl);
    if (!debug) expect(sdl).toMatchSnapshot();
  });
};

describe("types", () => {
  it("renders nothing for no changes", () => {
    const { current, next } = schemas(
      gql`
        type User {
          id: ID!
        }
      `,
      gql`
        type User {
          id: ID!
        }
      `
    );
    const sdl = printFromSchemas(current, next);
    expect(sdl).toBeFalsy();
  });

  compare(
    "shows removed types",
    gql`
      type RemovedType {
        id: ID!
      }

      enum RemovedEnum {
        ONE
      }

      input RemovedInput {
        id: ID!
      }

      scalar RemovedScalar

      interface RemovedInterface {
        id: ID!
      }

      union RemovedUnion = User

      type User {
        id: ID!
      }
    `,
    gql`
      type User {
        id: ID!
      }
    `
  );

  compare(
    "shows added types",
    gql`
      type User {
        id: ID!
      }
    `,
    gql`
      type AddedType {
        id: ID!
      }

      enum AddedEnum {
        ONE
      }

      input AddedInput {
        id: ID!
      }

      scalar AddedScalar

      interface AddedInterface {
        id: ID!
      }

      union AddedUnion = User

      type User {
        id: ID!
      }
    `
  );
});
describe("fields", () => {
  compare(
    "shows warning for removals",
    gql`
      type User {
        id: ID!
        firstName: String!
      }

      enum AddedEnum {
        ONE
        TWO
      }

      input AddedInput {
        id: ID!
        firstName: String!
        lastName: String
      }

      interface AddedInterface {
        id: ID!
        firstName(arg: String): String
      }
    `,
    gql`
      type User {
        id: ID!
      }

      enum AddedEnum {
        ONE
      }

      input AddedInput {
        id: ID!
      }

      interface AddedInterface {
        id: ID!
      }
    `
  );
  compare(
    "shows notice for additions",
    gql`
      type User {
        id: ID!
      }

      enum AddedEnum {
        ONE
      }

      input AddedInput {
        id: ID!
      }

      interface AddedInterface {
        id: ID!
      }
    `,
    gql`
      type User {
        id: ID!
        firstName: String!
      }

      enum AddedEnum {
        ONE
        TWO
      }

      input AddedInput {
        id: ID!
        firstName: String!
        lastName: String
      }

      interface AddedInterface {
        id: ID!
        firstName(arg: String): String
      }
    `
  );
});

describe("arguments", () => {
  compare(
    "shows removed arguments",
    gql`
      type Query {
        hello(name: String): String
      }

      interface Node {
        id(id: ID!): String
      }
    `,
    gql`
      type Query {
        hello: String
      }

      interface Node {
        id: String
      }
    `
  );
  compare(
    "shows added arguments",
    gql`
      type Query {
        hello: String
      }

      interface Node {
        id: String
      }
    `,
    gql`
      type Query {
        hello(name: String): String
      }

      interface Node {
        id(id: ID!): String
      }
    `
  );
});

describe("interfaces", () => {
  compare(
    "reports removal of interface implementations",
    gql`
      type Query {
        user(id: ID!): User
      }

      interface Node {
        id: ID!
      }

      type User implements Node {
        id: ID!
      }
    `,
    gql`
      type Query {
        user(id: ID!): User
      }

      interface Node {
        id: ID!
      }

      type User {
        id: ID!
      }
    `
  );
  compare(
    "reports removal multiple of interface implementations",
    gql`
      type Query {
        user(id: ID!): User
      }

      interface Node {
        id: ID!
      }

      interface Person {
        id: ID!
      }

      type User implements Node & Person {
        id: ID!
      }
    `,
    gql`
      type Query {
        user(id: ID!): User
      }

      interface Node {
        id: ID!
      }

      interface Person {
        id: ID!
      }

      type User implements Node {
        id: ID!
      }
    `
  );
  compare(
    "reports additiong of interface to object",
    gql`
      type Query {
        user(id: ID!): User
      }

      interface Node {
        id: ID!
      }

      type User {
        id: ID!
      }
    `,
    gql`
      type Query {
        user(id: ID!): User
      }

      interface Node {
        id: ID!
      }

      type User implements Node {
        id: ID!
      }
    `
  );

  compare(
    "reports adding multiple interface implementations",
    gql`
      type Query {
        user(id: ID!): User
      }

      interface Node {
        id: ID!
      }

      interface Person {
        id: ID!
      }

      type User implements Node {
        id: ID!
      }
    `,
    gql`
      type Query {
        user(id: ID!): User
      }

      interface Node {
        id: ID!
      }

      interface Person {
        id: ID!
      }

      type User implements Node & Person {
        id: ID!
      }
    `
  );
});

describe("unions", () => {
  compare(
    "type removed from union",
    gql`
      type User {
        id: ID!
      }

      type Person {
        firstName: String
      }

      union Client = User | Person
    `,
    gql`
      type User {
        id: ID!
      }

      type Person {
        firstName: String
      }

      union Client = User
    `
  );
  compare(
    "type added from union",
    gql`
      type User {
        id: ID!
      }

      type Person {
        firstName: String
      }

      union Client = User
    `,
    gql`
      type User {
        id: ID!
      }

      type Person {
        firstName: String
      }

      union Client = User | Person
    `
  );
});

describe("kind changes", () => {
  compare(
    "type changed kind",
    gql`
      type Query {
        hello: String
      }

      interface ChangeType {
        id: ID!
      }
    `,
    gql`
      type Query {
        hello: String
      }

      type ChangeType {
        id: ID!
      }
    `
  );
  compare(
    "field changed kind",
    gql`
      type Query {
        hello: String
      }

      input Hello {
        hello: String
      }
    `,
    gql`
      type Query {
        hello: ID!
      }

      input Hello {
        hello: String!
      }
    `
  );
  compare(
    "arg changed kind",
    gql`
      type Query {
        hello(id: ID): String
        defaultVal(id: String = "1"): String
      }
    `,
    gql`
      type Query {
        hello(id: String): String
        defaultVal(id: String = "2"): String
      }
    `
  );
});

describe("deprecation changes", () => {
  compare(
    "deprecation additions",
    gql`
      type Query {
        deprecated: String
        deprecatedWithReason: String @deprecated(reason: "Give up")
      }

      enum SoLong {
        TWO
        ONE @deprecated(reason: "There can only be one highlander")
      }
    `,
    gql`
      type Query {
        deprecated: String @deprecated
        deprecatedWithReason: String
          @deprecated(reason: "Use field testcase instead")
      }

      enum SoLong {
        TWO @deprecated
        ONE @deprecated(reason: "One is the loneliest number")
      }
    `
  );
  compare(
    "deprecation removals",
    gql`
      type Query {
        deprecated: String @deprecated
      }

      enum SoLong {
        TWO @deprecated
      }
    `,
    gql`
      type Query {
        deprecated: String
      }

      enum SoLong {
        TWO
      }
    `
  );
});

// // Post Apollo Day goals
// DIRECTIVE_REMOVED,
// DIRECTIVE_LOCATION_REMOVED,
// DIRECTIVE_ARG_REMOVED,
// NON_NULL_DIRECTIVE_ARG_ADDED,

describe("integration", () => {
  // XXX make this change complex
  it("reports changes for a complex scenario", () => {
    const { current, next } = schemas(initial, change);
    const changes = diffSchemas(current.getTypeMap(), next.getTypeMap());
    const sdl = printChanges(changes);
    expect(sdl).toMatchSnapshot();
  });
});
