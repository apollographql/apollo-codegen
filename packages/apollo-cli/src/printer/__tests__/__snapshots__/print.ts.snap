// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`arguments shows added arguments 1`] = `
"interface Node { 
  
  # ✔️ Notice ✔️
  id(id: ID!): String

}

type Query { 
  
  # ✔️ Notice ✔️
  hello(name: String): String

}

"
`;

exports[`arguments shows removed arguments 1`] = `
"
interface Node { 
  
  # ⚡ Warning ⚡
  id: String

}

type Query { 
  
  # ⚡ Warning ⚡
  hello: String

}

"
`;

exports[`deprecation changes deprecation additions 1`] = `
"enum SoLong { 
  
  # ⚡ Warning ⚡
  TWO @deprecated

  # ✔️ Notice ✔️
  ONE @deprecated(reason: \\"One is the loneliest number\\")

}

type Query { 
  
  # ⚡ Warning ⚡
  deprecated: String @deprecated

  # ✔️ Notice ✔️
  deprecatedWithReason: String @deprecated(reason: \\"Use field testcase instead\\")

}

"
`;

exports[`deprecation changes deprecation removals 1`] = `
"enum SoLong { 
  
  # ✔️ Notice ✔️
  TWO

}

type Query { 
  
  # ✔️ Notice ✔️
  deprecated: String

}

"
`;

exports[`fields shows notice for additions 1`] = `
"interface AddedInterface { 
  
  # ✔️ Notice ✔️
  firstName(arg: String): String

}

input AddedInput { 
  
  # ⚡ Warning ⚡
  firstName: String!

  # ✔️ Notice ✔️
  lastName: String

}

enum AddedEnum { 
  
  # ✔️ Notice ✔️
  TWO

}

type User { 
  
  # ✔️ Notice ✔️
  firstName: String!

}

"
`;

exports[`fields shows warning for removals 1`] = `
"interface AddedInterface { 
  
  # ⚡ Warning ⚡
  firstName(arg: String): String

}

input AddedInput { 
  
  # ⚡ Warning ⚡
  firstName: String!
  lastName: String

}

enum AddedEnum { 
  
  # ⚡ Warning ⚡
  TWO

}

type User { 
  
  # ⚡ Warning ⚡
  firstName: String!

}

"
`;

exports[`integration reports changes for a complex scenario 1`] = `
"type User { 
  
  # ✔️ Notice ✔️
  id: ID!
  firstName: String

}

type Query { 
  
  # ⚡ Warning ⚡
  hello: String!
  serverSideField: ServerField!

  # ✔️ Notice ✔️
  me: User

}

type RemovedType { }

type RemovedField { 
  
  # ⚡ Warning ⚡
  name: RemovedType

}

type ServerField { }

"
`;

exports[`interfaces reports adding multiple interface implementations 1`] = `
"type User implements Node & Person { }

"
`;

exports[`interfaces reports additiong of interface to object 1`] = `
"type User implements Node { }

"
`;

exports[`interfaces reports removal multiple of interface implementations 1`] = `
"type User implements Node { }

"
`;

exports[`interfaces reports removal of interface implementations 1`] = `
"type User { }

"
`;

exports[`kind changes arg changed kind 1`] = `
"
type Query { 
  
  # ⚡ Warning ⚡
  hello(id: String): String
  defaultVal(id: String = \\"2\\"): String

}

"
`;

exports[`kind changes field changed kind 1`] = `
"input Hello { 
  
  # ⚡ Warning ⚡
  hello: String!

}

type Query { 
  
  # ⚡ Warning ⚡
  hello: ID!

}

"
`;

exports[`kind changes type changed kind 1`] = `
"type ChangeType { }

"
`;

exports[`types shows added types 1`] = `
"union AddedUnion = User

interface AddedInterface { 
  
  # ✔️ Notice ✔️
  id: ID!

}

scalar AddedScalar

input AddedInput { 
  
  # ✔️ Notice ✔️
  id: ID!

}

enum AddedEnum { 
  
  # ✔️ Notice ✔️
  ONE

}

type AddedType { 
  
  # ✔️ Notice ✔️
  id: ID!

}

"
`;

exports[`types shows removed types 1`] = `
"union RemovedUnion = User

interface RemovedInterface { }

scalar RemovedScalar

input RemovedInput { }

enum RemovedEnum { }

type RemovedType { }

"
`;

exports[`unions type added from union 1`] = `
"union Client = User | Person

"
`;

exports[`unions type removed from union 1`] = `
"union Client = User

"
`;
